### PR TITLE
Keras Frontend

### PR DESCRIPTION
**Pull request by Huyuwei**
Now VGG16 pass the test.
Able to merge after xception and resnet50 pass the test. Will be soon. Please review then.

### CODE REVIEW

**Comment by tqchen** [comment]
cc @fchollet

**Comment by tqchen** [comment]
@zhreshold  can you help do a quick review?

**Comment by fchollet** [comment]
What's the high-level summary here?

File: python/nnvm/frontend/keras.py [hunk 1]
```diff
@@ -0,0 +1,367 @@
+# pylint: disable=invalid-name, import-self
+"""Keras frontend."""
+from __future__ import absolute_import as _abs
+import numpy as np
+import six
```
Comment (zhreshold) [inline python/nnvm/frontend/keras.py]: six is extra dependency

Comment (Huyuwei) [reply python/nnvm/frontend/keras.py]: ok, I will remove it.

**Comment by Huyuwei** [comment]
@fchollet 
It's to convert a keras model, e.g. keras.applications.vgg16.VGG16, to NNVM format. See the [coreml frontend tutorial](http://nnvm.tvmlang.org/tutorials/from_coreml.html#sphx-glr-tutorials-from-coreml-py) for a reference.

**Comment by tqchen** [comment]
The major goal is to quickly deploy Keras to bare metal device like rasp, web browser and accelerators

**Comment by fchollet** [comment]
Cool, thanks for looping me in. Any specific input you need from me?

**Comment by tqchen** [comment]
Yuwei will be putting up an tutorial for Keras compilation, which we could be using your input in reviewing the example

**Comment by Huyuwei** [comment]
This is ready to merge. Most operators can be converted correctly.

Resnet50 and xception fail due to asymmetry padding in certain conv ops. Maybe nnvm's conv2d should be modified to support asymmetry padding, i.e. to take padding params of length 4 [pad_top, pad_bottom, pad_left, pad_right].

**Comment by Huyuwei** [comment]
Now resnet50 pass test. Xception runs success with random weights and target='cuda', it fails with imagenet weights or when target='llvm'. I will look into this later.

**Comment by tqchen** [comment]
@Huyuwei any update on this?

**Comment by Huyuwei** [comment]
@tqchen Now xception is correctly converted and passes test with 'cuda' target. It fails with 'llvm' because llvm conv schedule doesn't work with 299x299x3 input size.

Another known bug is that keras and nnvm treat padded zeros in average pooling differently. Assume a 3x2 block on the boundary is padded to 3x3, nnvm divides the sum by 9 in calculating the average value, while keras divides the sum by 6.

**Comment by Huyuwei** [comment]
@tqchen Now all tests pass, after a bug in converting bias is fixed. Can be merged.

File: include/nnvm/top/nn.h [hunk 2]
```diff
@@ -101,6 +101,21 @@ struct LeakyReLUParam : public dmlc::Parameter<LeakyReLUParam> {
   }
 };
 
+
+struct PadParam : public dmlc::Parameter<PadParam> {
+  float pad_value;
+  Tuple<Tuple<int>> pad_width;
```
Comment (tqchen) [inline include/nnvm/top/nn.h]: use ```> >``` space between > to make it compatible with MSVC

**Comment by tqchen** [comment]
thanks! this is merged @Huyuwei can you followup with a tutorial PR?

**Comment by Huyuwei** [comment]
@tqchen Ok
